### ISSUE TITLE
Match clauses guaranteed tail call?

### ISSUE THREAD

**Issue by andrewchambers** [janet-lang/janet#341]
I have a function like so:
```
(defn recv
      []
      (match (protocol/recv-msg c)
        [:error msg]
          (error msg)
        [:stderr ln]
          (do
            (eprin ln)
            (recv))
        :sending-content
          (protocol/recv-file c destf)))
```
The recursive call to recv is not a tail call:
```
  in recv [src/fetch.janet] on line 120, column 11
  in recv [src/fetch.janet] on line 124, column 13
  in recv [src/fetch.janet] on line 124, column 13
  in recv [src/fetch.janet] on line 124, column 13
  in <anonymous> [src/fetch.janet] on line 127, column 5
  in <anonymous> [src/fetch.janet] on line 114, column 3
```

Should this be a tail call? When can we tell/assert a form will be a tailcall?

**Comment by bakpakin** [comment]
The way match is written, it will actually never be a tail call. We should be able to improve this, though, as there is no fundamental reason why it can't be a tail call.

**Comment by bakpakin** [comment]
This is addressed in 945cbcfad6ec1a59bb647cfe4c1d89e54efe4269

**Comment by pepe** [comment]
Oh sir you are the best!Thank you very much.
